cpp/src/lists/contains.cu

```diff
@@ -34,215 +34,182 @@ namespace lists {
-                        if (list.is_null()) { return false; }
-                        auto search_key = search_key_and_validity.first;
-
-                        return thrust::find_if(thrust::seq,
-                                               list.pair_begin<T>(),
-                                               list.pair_end<T>(),
-                                               [search_key] __device__(auto element_and_validity) {
-                                                 return element_and_validity.second &&
-                                                        (element_and_validity.first == search_key);
-                                               }) != list.pair_end<T>();
-                      });
+    if (!search_keys_have_nulls && !input_lists.has_nulls() && !input_lists.child().has_nulls()) {
+      return {rmm::device_buffer{0, stream, mr}, size_type{0}};
+    } else {
+      return cudf::detail::valid_if(result_validity.begin<bool>(),
+                                    result_validity.end<bool>(),
+                                    thrust::identity<bool>{},
+                                    stream,
+                                    mr);
+    }
   }
 
-  template <typename T>
-  std::enable_if_t<cudf::is_numeric<T>() || cudf::is_chrono<T>() ||
-                     std::is_same<T, cudf::string_view>::value,
-                   void>
-  operator()(
-    cudf::detail::lists_column_device_view const& d_lists,
-    cudf::scalar const& search_key,
-    bool search_key_is_null,  // Ignored for scalars. This is decipherable from `search_key`.
-    cudf::mutable_column_device_view output_bools,
-    rmm::cuda_stream_view stream,
-    rmm::mr::device_memory_resource* mr) const
+  template <typename ElementType, typename SearchKeyPairIter>
+  void search_each_list_row(cudf::detail::lists_column_device_view const& d_lists,
+                            SearchKeyPairIter search_key_pair_iter,
+                            cudf::mutable_column_device_view mutable_ret_bools,
+                            cudf::mutable_column_device_view mutable_ret_validity,
+                            rmm::cuda_stream_view stream,
+                            rmm::mr::device_memory_resource* mr)
   {
-    assert(search_key.is_valid() && "skey should have been checked for nulls by this point.");
-    return op_impl<T>(d_lists, detail::make_pair_iterator<T>(search_key), output_bools, stream, mr);
+    thrust::for_each(
+      rmm::exec_policy(stream),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(d_lists.size()),
+      [d_lists,
+       search_key_pair_iter,
+       d_bools    = mutable_ret_bools.data<bool>(),
+       d_validity = mutable_ret_validity.data<bool>()] __device__(auto row_index) {
+        auto search_key_and_validity    = search_key_pair_iter[row_index];
+        auto const& search_key_is_valid = search_key_and_validity.second;
```

Comment:
Doing a for-each over a counting iterator used to index into a pair iterator kind of defeats the benefit of a pair iterator....

Whenever you have an algorithm with a lambda that reads and writes to the same explicit location (`[row_index]`), it's a hint that you are not using iterator machinery to its strengths.

Why not use the pair iterator begin and end as the range and then get the pair as the argument to the lambda?

Also, don't we have an output pair iterator so that you can just `return thrust::pair<false, false>` (for example) rather than assigning to two different arrays (at the same index in each code path).  Then you can replace the `for_each` with a `transform`.

Reply:
> Whenever you have an algorithm with a lambda that reads and writes to the same explicit location ([row_index]), it's a hint that you are not using iterator machinery to its strengths.

Thanks for pointing that out, @harrism. The issue was line#121:
```c++
auto list = cudf::list_device_view(d_lists, row_index);
```
I need the `row_index` to construct the `list_device_view` for that list row. I could not figure out a way around that. :/

Reply:
>  don't we have an output pair iterator ...?

Sorry, do we have a `pair_begin()` on `mutable_column_device_view`? I'm afraid I can't seem to locate it.

Reply:
Isn't there an iterator for that? The binary form of `thrust::transform` would allow you to iterate over the search keys and the lists column together...

Reply:
Messages crossed. I'm looking...

Reply:
Why does `list` need to be mutable? Seems to be read-only?

Reply:
There is no output pair iterator for `mutable_column_device_view`. Such an iterator would be extremely inefficient as writing to the bitmask via the iterator would incur atomics for each bit modified.

Reply:
> Why does `list` need to be mutable? Seems to be read-only?

`list` is not mutable. `list` is a `list_device_view`, akin to `string_view`. 
Sorry, I am unable to follow how the `list` mutability affects its construction from `lists_column_device_view` and `row_index`.